store/tikv/client.go

```diff
@@ -172,6 +292,126 @@ func (a *connArray) Close() {
 	close(a.streamTimeout)
 }
 
+type batchCommandsEntry struct {
+	req     *tikvpb.BatchCommandsRequest_Request
+	res     chan *tikvpb.BatchCommandsResponse_Response
+	timeout int32 // Indicates the request is timeout or not.
+	err     error
+}
+
+func (a *connArray) batchSendLoop(cfg config.TiKVClient) {
+	entries := make([]*batchCommandsEntry, 0, cfg.MaxBatchSize)
+	requests := make([]*tikvpb.BatchCommandsRequest_Request, 0, cfg.MaxBatchSize)
+	requestIDs := make([]uint64, 0, cfg.MaxBatchSize)
+
+	for {
+		metrics.TiKVPendingBatchRequests.Set(float64(len(a.batchCommandsCh)))
+
+		// Choose a connection by round-robbin.
+		next := atomic.AddUint32(&a.index, 1) % uint32(len(a.v))
+		batchCommandsClient := a.batchCommandsClients[next]
+
+		tikvTransportLayerLoad := atomic.LoadUint64(batchCommandsClient.tikvTransportLayerLoad)
+		inHeavyLoad := uint(tikvTransportLayerLoad) >= cfg.TiKVHeavyLoadToBatch // Need to wait.
+
+		entries = entries[:0]
+		requests = requests[:0]
+		requestIDs = requestIDs[:0]
+
+		// Block on the first element.
+		headEntry := <-a.batchCommandsCh
+		if headEntry == nil {
+			return
+		}
+		entries = append(entries, headEntry)
+
+		// This loop is for trying best to collect more requests.
+	Loop:
+		for {
+			select {
+			case entry := <-a.batchCommandsCh:
+				if entry == nil {
+					return
+				}
+				entries = append(entries, entry)
+				if len(entries) >= int(cfg.MaxBatchSize) {
+					break Loop
+				}
+			default:
+				break Loop
+			}
+		}
+
+		// If the target TiKV is overload, wait a while to collect more requests.
+		if len(entries) < int(cfg.MaxBatchSize) && inHeavyLoad && cfg.MaxBatchWaitTime > 0 {
+			waitStart := time.Now()
+			var waitEnd time.Time
+			after := time.After(cfg.MaxBatchWaitTime)
+			batchIsFilled := false
```

Comment:
```suggestion
			var batchIsFilled bool
```
@rleungx is adding `golangci-lint` to TiDB, and this may introduce a warning.